[expo-audio][android] Unable to initialize audio player with null source

### Minimal reproducible example

https://github.com/fit52-digital/video-library-testing

### What platform(s) does this occur on?

Android

### Where did you reproduce the issue?

in a development build

### Summary

I found that initializing the audio player with a null source (as I intended to re-use the audio player) causes a crash on android.
This does not occur on iOS.

The `useAudioPlayer` accepts `null` as a source, so this seems to be a bug.

```typescript
export function useAudioPlayer(
  source: AudioSource = null,
  ```


```typescript
const ExpoAudioPlayerListItem: React.FC<IExpoAudioPlayerItemProps> = props => {
  const {source, isPlaying = false, index} = props;

  const audioPlayer = useAudioPlayer(null);

  useEffect(() => {
    if (audioPlayer) {
      if (!isPlaying) {
        audioPlayer.pause();
      } else {
        audioPlayer.play();
      }
    }
  }, [isPlaying, audioPlayer]);

  useEffect(() => {
    if (audioPlayer) {
      audioPlayer.replace(source);
    }
  }, [source, audioPlayer]);
  ```


Error:
```
Warning: Error: Call to function 'AudioPlayer.constructor' has been rejected.
→ Caused by: java.lang.NullPointerException: uriString

This error is located at:
    in ExpoAudioPlayerListItem (created by ExpoAudioScreenTest1)
    in RCTView (created by View)
    in View (created by ScrollView)
    in RCTScrollView
    in VScrollViewNativeComponent (created by ScrollView)
    in ScrollView (created by ScrollView)
    in Wrapper (created by ExpoAudioScreenTest1)
    in RCTView (created by View)
    in View (created by ExpoAudioScreenTest1)
    in ExpoAudioScreenTest1 (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by NativeStackNavigator)
    in RNSScreenContentWrapper (created by ScreenContentWrapper)
    in ScreenContentWrapper (created by DebugContainer)
    in DebugContainer
    in RNSScreen (created by Animated(Anonymous))
    in Animated(Anonymous)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze
    in InnerScreen (created by Screen)
    in Screen
    in ScreenStackItem (created by SceneView)
    in SceneView (created by NativeStackView)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze (created by ScreenStack)
    in RNSScreenStack (created by ScreenStack)
    in Unknown (created by ScreenStack)
    in ScreenStack (created by NativeStackView)
    in RCTView (created by View)
    in View (created by SafeAreaProviderCompat)
    in SafeAreaProviderCompat (created by NativeStackView)
    in NativeStackView (created by NativeStackNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by NativeStackNavigator)
    in NativeStackNavigator (created by HomeStackNavigator)
    in HomeStackNavigator (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by BottomTabNavigator)
    in RCTView (created by View)
    in View (created by Screen)
    in RCTView (created by View)
    in View (created by Animated(View))
    in Animated(View) (created by Background)
    in Background (created by Screen)
    in Screen (created by BottomTabView)
    in RNSScreen (created by Animated(Anonymous))
    in Animated(Anonymous)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze
    in InnerScreen (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by BottomTabView)
    in RNSScreenContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by BottomTabView)
    in RCTView (created by View)
    in View (created by SafeAreaProviderCompat)
    in SafeAreaProviderCompat (created by BottomTabView)
    in BottomTabView (created by BottomTabNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by BottomTabNavigator)
    in BottomTabNavigator (created by Tabs)
    in Tabs (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by NativeStackNavigator)
    in RNSScreenContentWrapper (created by ScreenContentWrapper)
    in ScreenContentWrapper (created by DebugContainer)
    in DebugContainer
    in RNSScreen (created by Animated(Anonymous))
    in Animated(Anonymous)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze
    in InnerScreen (created by Screen)
    in Screen
    in ScreenStackItem (created by SceneView)
    in SceneView (created by NativeStackView)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze (created by ScreenStack)
    in RNSScreenStack (created by ScreenStack)
    in Unknown (created by ScreenStack)
    in ScreenStack (created by NativeStackView)
    in RNCSafeAreaProvider (created by SafeAreaProvider)
    in SafeAreaProvider (created by SafeAreaProviderCompat)
    in SafeAreaProviderCompat (created by NativeStackView)
    in NativeStackView (created by NativeStackNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by NativeStackNavigator)
    in NativeStackNavigator (created by RootNavigator)
    in ThemeProvider
    in EnsureSingleNavigator
    in BaseNavigationContainer
    in NavigationContainerInner (created by RootNavigator)
    in RootNavigator (created by App)
    in App
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in VideoLibraryTesting(RootComponent)
```

Hope this helps 





### Environment

```text
expo-env-info 1.2.1 environment info:
    System:
      OS: macOS 15.1.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.5.1 - /opt/homebrew/bin/node
      Yarn: 1.22.22 - /opt/homebrew/bin/yarn
      npm: 10.8.2 - /opt/homebrew/bin/npm
      Watchman: 2024.07.15.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /Users/mihailapuste/.rvm/gems/ruby-3.1.2/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.1, iOS 18.1, macOS 15.1, tvOS 18.1, visionOS 2.1, watchOS 11.1
    IDEs:
      Android Studio: 2024.2 AI-242.23339.11.2421.12550806
      Xcode: 16.1/16B40 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^52.0.0 => 52.0.20 
      react: 18.3.1 => 18.3.1 
      react-native: 0.76.5 => 0.76.5 
    npmGlobalPackages:
      eas-cli: 7.1.3
    Expo Workflow: bare
```


### Expo Doctor Diagnostics

```text
Multiple lock files detected (yarn.lock, package-lock.json). This may result in unexpected behavior in CI environments, such as EAS Build, which infer the package manager from the lock file.

Error: Problem validating fields in /Users/mihailapuste/VideoLibraryTesting/app.json. Learn more: https://docs.expo.dev/workflow/configuration/
 • should NOT have additional property 'displayName'.

The following issues were found when validating your dependencies against React Native Directory:
No metadata available: @emotion/core, @emotion/native, expo-audio, expo-video, mobx, mobx-react-lite, mobx-state-tree
Advice: 
- Update React Native Directory to include metadata for unknown packages. Alternatively, set expo.doctor.reactNativeDirectoryCheck.listUnknownPackages in package.json to false to skip warnings about packages with no metadata, if the warning is not relevant.

The following packages should be updated for best compatibility with the installed expo version:
  react-native-safe-area-context@5.0.0 - expected version: 4.12.0
  react-native-screens@4.3.0 - expected version: ~4.1.0
  react-native-svg@15.10.1 - expected version: 15.8.0
  typescript@5.0.4 - expected version: ^5.3.3
Your project may not work correctly until you install the expected versions of the packages.
Found outdated dependencies
Advice: Use 'npx expo install --check' to review and upgrade your dependencies.
```

--- a/packages/expo-audio/CHANGELOG.md
+++ b/packages/expo-audio/CHANGELOG.md
@@ -11,6 +11,7 @@
 ### 🐛 Bug fixes
 
 - Expose `isMeteringEnabled` to `JS`. ([#33713](https://github.com/expo/expo/pull/33713) by [@alanjhughes](https://github.com/alanjhughes))
+- On `Android`, allow player to accept a `null` audio source.
 
 ### 💡 Others
 

--- a/packages/expo-audio/CHANGELOG.md
+++ b/packages/expo-audio/CHANGELOG.md
@@ -11,7 +11,7 @@
 ### 🐛 Bug fixes
 
 - Expose `isMeteringEnabled` to `JS`. ([#33713](https://github.com/expo/expo/pull/33713) by [@alanjhughes](https://github.com/alanjhughes))
-- On `Android`, allow player to accept a `null` audio source.
+- On `Android`, allow player to accept a `null` audio source. ([#33854](https://github.com/expo/expo/pull/33854) by [@alanjhughes](https://github.com/alanjhughes))
 
 ### 💡 Others
 

--- a/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioModule.kt
+++ b/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioModule.kt
@@ -239,7 +239,7 @@ class AudioModule : Module() {
           ref.player.volume
         }
       }.set { ref, volume: Float ->
-        appContext.mainQueue.launch {
+        runOnMain {
           ref.player.volume = volume
         }
       }
@@ -249,21 +249,23 @@ class AudioModule : Module() {
           Log.e(TAG, "Audio has been disabled. Re-enable to start playing")
           return@Function
         }
-        appContext.mainQueue.launch {
+        runOnMain {
           ref.player.play()
         }
       }
 
       Function("pause") { ref: AudioPlayer ->
-        appContext.mainQueue.launch {
+        runOnMain {
           ref.player.pause()
         }
       }
 
-      Function("replace") { ref: AudioPlayer, source: AudioSource ->
+      Function("replace") { ref: AudioPlayer, source: AudioSource? ->
         if (ref.player.availableCommands.contains(Player.COMMAND_CHANGE_MEDIA_ITEMS)) {
           val mediaSource = createMediaItem(source)
-          ref.player.replaceMediaItem(0, mediaSource.mediaItem)
+          mediaSource?.let {
+            ref.player.replaceMediaItem(0, it.mediaItem)
+          }
         }
       }
 
@@ -355,21 +357,26 @@ class AudioModule : Module() {
     }
   }
 
-  private fun createMediaItem(source: AudioSource?): MediaSource {
-    val isLocal = Util.isLocalFileUri(Uri.parse(source?.uri))
-    val factory = if (isLocal) {
+  private fun createMediaItem(source: AudioSource?): MediaSource? = source?.let {
+    val factory = createDataSourceFactory(it)
+    it.uri?.let { uri ->
+      val item = MediaItem.fromUri(uri)
+      buildMediaSourceFactory(factory, item)
+    }
+  }
+
+  private fun createDataSourceFactory(audioSource: AudioSource): DataSource.Factory {
+    val isLocal = Util.isLocalFileUri(Uri.parse(audioSource.uri))
+    return if (isLocal) {
       DefaultDataSource.Factory(context)
     } else {
       OkHttpDataSource.Factory(httpClient).apply {
-        source?.headers?.let {
-          setDefaultRequestProperties(it)
+        audioSource.headers?.let { headers ->
+          setDefaultRequestProperties(headers)
         }
         DefaultDataSource.Factory(context, this)
       }
     }
-
-    val item = MediaItem.fromUri(source?.uri ?: "")
-    return buildMediaSourceFactory(factory, item)
   }
 
   private fun updatePlaySoundThroughEarpiece(playThroughEarpiece: Boolean) {
@@ -384,7 +391,7 @@ class AudioModule : Module() {
     mediaItem: MediaItem
   ): MediaSource {
     val uri = mediaItem.localConfiguration?.uri
-    val newFactory = when (val type = retrieveStreamType(uri!!)) {
+    val newFactory = when (val type = uri?.let { retrieveStreamType(it) }) {
       CONTENT_TYPE_SS -> SsMediaSource.Factory(factory)
       CONTENT_TYPE_DASH -> DashMediaSource.Factory(factory)
       CONTENT_TYPE_HLS -> HlsMediaSource.Factory(factory)

--- a/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioPlayer.kt
+++ b/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioPlayer.kt
@@ -7,6 +7,7 @@ import android.media.audiofx.Visualizer
 import android.util.Log
 import androidx.core.content.ContextCompat
 import androidx.media3.common.AudioAttributes
+import androidx.media3.common.C
 import androidx.media3.common.Player
 import androidx.media3.common.util.UnstableApi
 import androidx.media3.exoplayer.ExoPlayer
@@ -29,14 +30,16 @@ private const val AUDIO_SAMPLE_UPDATE = "audioSampleUpdate"
 class AudioPlayer(
   context: Context,
   appContext: AppContext,
-  source: MediaSource,
+  source: MediaSource?,
   updateInterval: Double
 ) : SharedRef<ExoPlayer>(
   ExoPlayer.Builder(context)
     .setLooper(context.mainLooper)
     .build()
     .apply {
-      setMediaSource(source)
+      source?.let {
+        setMediaSource(it)
+      }
       setAudioAttributes(AudioAttributes.DEFAULT, true)
       prepare()
     },
@@ -52,7 +55,7 @@ class AudioPlayer(
   private var visualizer: Visualizer? = null
 
   val currentTime get() = player.currentPosition / 1000
-  val duration get() = player.duration / 1000
+  val duration get() = if (player.duration != C.TIME_UNSET) player.duration / 1000 else 0
 
   init {
     addPlayerListeners()
@@ -175,7 +178,8 @@ class AudioPlayer(
               }
             }
 
-            override fun onFftDataCapture(visualizer: Visualizer?, fft: ByteArray?, samplingRate: Int) = Unit
+            override fun onFftDataCapture(visualizer: Visualizer?, fft: ByteArray?, samplingRate: Int) =
+              Unit
           },
           Visualizer.getMaxCaptureRate() / 2,
           true,

--- a/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioPlayer.kt
+++ b/packages/expo-audio/android/src/main/java/expo/modules/audio/AudioPlayer.kt
@@ -178,8 +178,7 @@ class AudioPlayer(
               }
             }
 
-            override fun onFftDataCapture(visualizer: Visualizer?, fft: ByteArray?, samplingRate: Int) =
-              Unit
+            override fun onFftDataCapture(visualizer: Visualizer?, fft: ByteArray?, samplingRate: Int) = Unit
           },
           Visualizer.getMaxCaptureRate() / 2,
           true,

